service.go

```diff
@@ -69,25 +151,255 @@ func (c *controller) rmServiceBinding(name, sid, nid, eid string, ip net.IP) err
 		c.Unlock()
 		return nil
 	}
+	c.Unlock()
 
-	netBackEnds, ok := s.backEnds[nid]
+	s.Lock()
+	lb, ok := s.loadBalancers[nid]
 	if !ok {
-		c.Unlock()
+		s.Unlock()
 		return nil
 	}
 
-	delete(netBackEnds, eid)
+	// Delete the special "tasks.svc_name" backend record.
+	n.(*network).deleteSvcRecords("tasks."+name, ip, nil, false)
+	delete(lb.backEnds, eid)
+
+	if len(lb.backEnds) == 0 {
+		// All the backends for this service have been
+		// removed. Time to remove the load balancer and also
+		// remove the service entry in IPVS.
+		rmService = true
+
+		// Make sure to remove the right IP since if vip is
+		// not valid we would have added a DNS RR record.
+		svcIP := vip
+		if len(svcIP) == 0 {
+			svcIP = ip
+		}
 
-	if len(netBackEnds) == 0 {
-		delete(s.backEnds, nid)
+		n.(*network).deleteSvcRecords(name, svcIP, nil, false)
+		delete(s.loadBalancers, nid)
 	}
 
-	if len(s.backEnds) == 0 {
+	if len(s.loadBalancers) == 0 {
+		// All loadbalancers for the service removed. Time to
+		// remove the service itself.
 		delete(c.serviceBindings, sid)
 	}
-	c.Unlock()
+	s.Unlock()
+
+	// Remove loadbalancer service(if needed) and backend in all
+	// sandboxes in the network only if the vip is valid.
+	if len(vip) != 0 {
+		n.(*network).rmLBBackend(ip, vip, lb.fwMark, rmService)
+	}
+
+	return nil
+}
+
+// Get all loadbalancers on this network that is currently discovered
+// on this node..
+func (n *network) connectedLoadbalancers() []*loadBalancer {
+	c := n.getController()
+
+	c.Lock()
+	defer c.Unlock()
+
+	var lbs []*loadBalancer
+	for _, s := range c.serviceBindings {
+		if lb, ok := s.loadBalancers[n.ID()]; ok {
+			lbs = append(lbs, lb)
+		}
+	}
+
+	return lbs
+}
+
+// Populate all loadbalancers on the network that the passed endpoint
+// belongs to, into this sandbox.
+func (sb *sandbox) populateLoadbalancers(ep *endpoint) {
+	n := ep.getNetwork()
+	for _, lb := range n.connectedLoadbalancers() {
+		// Skip if vip is not valid.
+		if len(lb.vip) == 0 {
+			continue
+		}
+
+		addService := true
+		for _, ip := range lb.backEnds {
+			sb.addLBBackend(ip, lb.vip, lb.fwMark, addService)
+			addService = false
+		}
+	}
+}
+
+// Add loadbalancer backend to all sandboxes which has a connection to
+// this network. If needed add the service as well, as specified by
+// the addService bool.
+func (n *network) addLBBackend(ip, vip net.IP, fwMark uint32, addService bool) {
+	n.WalkEndpoints(func(e Endpoint) bool {
+		ep := e.(*endpoint)
+		sb, ok := ep.getSandbox()
```

Comment:
Minor: 
Given the walk function will never interrupt the walk, we could rewrite it as follow, it also seems more in line with the goal

```
func(e Endpoint) bool {
    ep := e.(*endpoint)
    if sb, ok := ep.getSandbox(); ok {
        sb.addLBBackend(ip, vip, fwMark, addService)
    }
    return false
}
```
Same in `rmLBBackend()`